File: src/omnia/data/structure/tuple/ImmutableCouplet.java
package omnia.data.structure.tuple;

import java.util.Iterator;
import java.util.OptionalInt;
import java.util.stream.Stream;
import omnia.data.structure.immutable.ImmutableList;

class ImmutableCouplet<T> extends ImmutableCouple<T, T> implements Couplet<T> {

  ImmutableCouplet(T first, T second) {
    super(first, second);
  }

  @Override
  public Triplet<T> concat(T object) {
    return Tuplet.of(first(), second(), object);
  }

  @Override
  public Quadruplet<T> concat(Couple<T, T> other) {
    return Tuplet.of(first(), second(), other.first(), other.second());
  }

  @Override
  public Quintuplet<T> concat(Triple<T, T, T> other) {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third());
  }

  @Override
  public Sextuplet<T> concat(Quadruple<T, T, T, T> other) {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third(), other.fourth());
  }

  @Override
  public Septuplet<T> concat(Quintuple<T, T, T, T, T> other) {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third(), other.fourth(), other.fifth());
  }

  @Override
  public Octuplet<T> concat(Sextuple<T, T, T, T, T, T> other) {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third(), other.fourth(), other.fifth(), other.sixth());
  }

  @Override
  public Nonuplet<T> concat(Septuple<T, T, T, T, T, T, T> other) {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third(), other.fourth(), other.fifth(), other.sixth(), other.seventh());
  }

  @Override
  public Decuplet<T> concat(Octuple<T, T, T, T, T, T, T, T> other) {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third(), other.fourth(), other.fifth(), other.sixth(), other.seventh(), other.eighth());
  }

  @Override
  public Iterator<T> iterator() {
    return toActualList().iterator();
  }

  @Override
  public T itemAt(int index) {
    return toActualList().itemAt(index);
  }

  @Override
  public OptionalInt indexOf(Object item) {
    return toActualList().indexOf(item);
  }

  @Override
  public Stream<T> stream() {
    return toActualList().stream();
  }

  @Override
  public boolean containsUnknownTyped(Object element) {
    return toActualList().containsUnknownTyped(element);
  }

  private ImmutableList<T> toActualList() {
    return ImmutableList.of(first(), second());
  }
}

File: src/omnia/data/structure/tuple/ImmutableCouplet.kt
package omnia.data.structure.tuple

import omnia.data.structure.immutable.ImmutableList
import java.util.OptionalInt
import java.util.function.Function
import java.util.stream.Stream

internal class ImmutableCouplet<T>(first: T, second: T) : ImmutableCouple<T, T>(first, second), Couplet<T> {
  override fun <R> map(mapper: Function<in T, out R>): Couplet<R> {
    return Tuplet.of(mapper.apply(first()), mapper.apply(second()))
  }

  override fun concat(`object`: T): Triplet<T> {
    return Tuplet.of(first(), second(), `object`)
  }

  override fun concat(other: Couple<T, T>): Quadruplet<T> {
    return Tuplet.of(first(), second(), other.first(), other.second())
  }

  override fun concat(other: Triple<T, T, T>): Quintuplet<T> {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third())
  }

  override fun concat(other: Quadruple<T, T, T, T>): Sextuplet<T> {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third(), other.fourth())
  }

  override fun concat(other: Quintuple<T, T, T, T, T>): Septuplet<T> {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third(), other.fourth(), other.fifth())
  }

  override fun concat(other: Sextuple<T, T, T, T, T, T>): Octuplet<T> {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third(), other.fourth(), other.fifth(), other.sixth())
  }

  override fun concat(other: Septuple<T, T, T, T, T, T, T>): Nonuplet<T> {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third(), other.fourth(), other.fifth(), other.sixth(), other.seventh())
  }

  override fun concat(other: Octuple<T, T, T, T, T, T, T, T>): Decuplet<T> {
    return Tuplet.of(first(), second(), other.first(), other.second(), other.third(), other.fourth(), other.fifth(), other.sixth(), other.seventh(), other.eighth())
  }

  override fun iterator(): Iterator<T> {
    return toActualList().iterator()
  }

  override fun itemAt(index: Int): T {
    return toActualList().itemAt(index)
  }

  override fun indexOf(item: Any?): OptionalInt {
    return toActualList().indexOf(item)
  }

  override fun stream(): Stream<T> {
    return toActualList().stream()
  }

  override fun containsUnknownTyped(item: Any?): Boolean {
    return toActualList().containsUnknownTyped(item)
  }

  private fun toActualList(): ImmutableList<T> {
    return ImmutableList.of(first(), second())
  }
}